### ISSUE TITLE
Image::decompress() does not seem to do anything.

### ISSUE THREAD

**Issue by Goblinth** [godotengine/godot#51528]
### Godot version

3.3.2 stable.mono  (I also tested 3.1.1 stable -> 3.3.1 stable) (not mono)

### System information

Windows 10, GLES3, GTX 1070ti

### Issue description

Image::compress()
ETC2 does not seem to do anything for jpg or png.
ETC does not seem to do anything for jpg, png changes to format 6 (causing is_compressed to return false)

Image::decompress()
Does not seem to do anything for jpg or png on any compression format (besides change the format back to what it was originally). Visually the image remains compressed.

I could be misunderstanding what decompress means in this context, but the docs likely need to be updated if so.

### Steps to reproduce

Load an Image with Image::load()
Compress the Image with Image::compress()
Attempt to decompress the Image with Image::decompress()

### Minimal reproduction project

[decompress_test.zip](https://github.com/godotengine/godot/files/6971641/decompress_test.zip)

**Comment by Calinou** [comment]
Check the return value of the `decompress` method: https://github.com/Calinou/godot/blob/61f860bc4b1ed34a8991a241cdd5f0d54f1c0cfe/core/io/image.cpp#L2365-L2380

**Comment by Goblinth** [comment]
S3TC returns 0, the others return ERR_UNAVAILABLE.
load_png_from_buffer() had the same result
They are being loaded into memory? Or am I misunderstanding something.

**Comment by Calinou** [comment]
> load_png_from_buffer() had the same result
They are being loaded into memory? Or am I misunderstanding something.

`load_png_from_buffer()` is designed to load binary data into an Image instance. It doesn't expect a file path. See the Image class documentation for an example of run-time image loading.

**Comment by Goblinth** [comment]
Yes, the code I used was
    `var f:File = File.new()
	f.open(IMAGE_TEST_PATH, File.READ)
	var e:int = i.load_png_from_buffer(f.get_buffer(f.get_len()))
	f.close()`

**Comment by Zylann** [comment]
When you load a PNG or JPG, with `load` or `load_png_from_buffer`, it decompresses the format right away because these are *file* formats, so what ends up in memory is the decompressed data already. So `decompress` does not do anything further because there is nothing to decompress at this point.

Formats like ETC or RGTC are intented for VRAM usage, so when you load them, they are not decompressed, because this will only happen on the graphics card.  `decompress` exists so that you can still decompress these images on the CPU to read their pixels there, that's why it only has an effect on VRAM formats.

**Comment by Goblinth** [comment]
Yes, I am trying to use decompress() on images that have already been compressed using Godot's compress() function. decompress() fails with error 2 for every format except S3TC (though it still changes the format number and is_compressed() will return false afterwards). S3TC is the only one to return 0, and the image remains the same visually. 

If I am understanding your second point correctly, then decompress() just makes it so that you can access/edit the pixels of the compressed image? My understanding from their names and the limited info in the docs was that decompress() reverted compress(), but this does not seem to be the case.

**Comment by akien-mga** [comment]
See the documentation for `Image.decompress()`:

https://github.com/godotengine/godot/blob/07c355c209c898f0696983f16cb2076366e0b283/doc/classes/Image.xml#L134-L140

So depending on which VRAM compression formats you're trying to decompress, it's possibly indeed that only S3TC works.

**Comment by Goblinth** [comment]
Here are the results of each compression format: 

Compression Format: S3TC
	Before Compression:
		compressed: False
		format: 5
	After Compression:		# changed visually
		compressed: True
		format: 19
	After Decompression:	# did not change visually
		error: 0
		compressed: False
		format: 5


Compression Format: PVRTC2
	Before Compression:
		compressed: False
		format: 5
	After Compression:		# changed visually
		compressed: True
		format: 28
	After Decompression:	# did not change visually
		error: 2
		compressed: False
		format: 5


Compression Format: PVRTC4
	Before Compression:
		compressed: False
		format: 5
	After Compression:		# changed visually
		compressed: True
		format: 28
	After Decompression:	# did not change visually
		error: 2
		compressed: False
		format: 5

Compression Format: ETC
	Before Compression:
		compressed: False
		format: 5
	After Compression:		# changed visually
		compressed: False
		format: 6
	After Decompression:	# did not change visually
		error: 2
		compressed: False
		format: 6


Compression Format: ETC2
	Before Compression:
		compressed: False
		format: 5
	After Compression:		# did not change visually
		compressed: True	
		format: 35
	After Decompression:	# did not change visually
		error: 2
		compressed: False
		format: 5

**Comment by akien-mga** [comment]
So to clarify, decompression is not just undo after compression, it needs a dedicated implementation that knows how to decompress the format (GPU drivers know to, but here it's software side).

We have function pointers that make it possible to implement decompression for each VRAM compressed format:
```
core/image.h
144:    static void (*_image_decompress_pvrtc)(Image *);
145:    static void (*_image_decompress_bc)(Image *);
146:    static void (*_image_decompress_bptc)(Image *);
147:    static void (*_image_decompress_etc1)(Image *);
148:    static void (*_image_decompress_etc2)(Image *);
```

The `_image_decompress_etc1` and `_image_decompress_etc2` functions are not implemented anywhere.

The others are implemented in their respective modules:
```
modules/cvtt/register_types.cpp
40:     Image::_image_decompress_bptc = image_decompress_cvtt;

modules/squish/register_types.cpp
37:     Image::_image_decompress_bc = image_decompress_squish;

modules/pvr/texture_loader_pvr.cpp
188:    Image::_image_decompress_pvrtc = _pvrtc_decompress;
```

PVRTC support is very basic (and partly broken, it seems to have PVRTC2 and PVRTC4 the same), so it might well be that the decompress method doesn't work.

**Comment by Goblinth** [comment]
As a final bit of clarification, is decompress() supposed to change an image visually, or just allow the pixels of the compressed image to be accessed?

**Comment by Calinou** [comment]
> As a final bit of clarification, is decompress() supposed to change an image visually, or just allow the pixels of the compressed image to be accessed?

It shouldn't change an image visually, as it's not possible to "reverse" lossy compression.
